File: src/main/java/pl/yalgrin/playnite/simplesync/web/objects/SourceResource.java
package pl.yalgrin.playnite.simplesync.web.objects;

import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.*;
import pl.yalgrin.playnite.simplesync.dto.objects.SourceDTO;
import pl.yalgrin.playnite.simplesync.helper.SingleExecutorHelper;
import pl.yalgrin.playnite.simplesync.service.objects.SourceService;
import reactor.core.publisher.Mono;

@RestController
@RequestMapping("/api/source")
@RequiredArgsConstructor
public class SourceResource {
    private final SourceService service;
    private final SingleExecutorHelper singleExecutorHelper;

    @GetMapping("/{id}")
    public Mono<SourceDTO> getSource(@PathVariable Long id) {
        return service.findById(id);
    }

    @PostMapping("/save")
    public Mono<SourceDTO> saveSource(@RequestBody SourceDTO dto, @RequestParam String clientId) {
        return singleExecutorHelper.runOnExecutor(service.saveObject(dto, clientId));
    }

    @PostMapping("/delete")
    public Mono<Void> deleteSource(@RequestBody SourceDTO dto, @RequestParam String clientId) {
        return singleExecutorHelper.runOnExecutor(service.deleteObject(dto, clientId));
    }
}

File: src/main/kotlin/pl/yalgrin/playnite/simplesync/web/library/SourceResource.kt
package pl.yalgrin.playnite.simplesync.web.library

import org.springframework.web.bind.annotation.*
import pl.yalgrin.playnite.simplesync.helper.SingleExecutorHelper
import pl.yalgrin.playnite.simplesync.library.dto.SourceDTO
import pl.yalgrin.playnite.simplesync.library.service.SourceService
import reactor.core.publisher.Mono

@RestController
@RequestMapping("/api/source")
class SourceResource(
    private val service: SourceService,
    private val singleExecutorHelper: SingleExecutorHelper
) {

    @GetMapping("/{id}")
    fun getSource(@PathVariable id: Long): Mono<SourceDTO> {
        return service.findById(id)
    }

    @PostMapping("/save")
    fun saveSource(@RequestBody dto: SourceDTO): Mono<SourceDTO> {
        return singleExecutorHelper.runOnExecutor(service.saveObjectAndPublishChanges(dto))
    }

    @PostMapping("/delete")
    fun deleteSource(@RequestBody dto: SourceDTO): Mono<Unit> {
        return singleExecutorHelper.runOnExecutor(service.deleteObjectAndPublishChanges(dto))
    }
}